### DISCUSSION TITLE
sample weights

### DISCUSSION THREAD

**Discussion by gcasamat** [DoubleML/doubleml-for-py#219] [Q&A] [answered]
I want to use RanfomForestRegressor as learners in the DoubleMLPLR procedure. Moreover I want to use sample weights for these learners. In scikit learn, It is possible by passing a vector of weights as follows: RandomForestRegressor().fit(X, y, sample_weight=weights).
I wonder if it is possible to pass such weights in RandomForestRegressor when using DoubleMLPLR?

**Comment by SvenKlaassen** [comment]
Currently this is not possible as we perform multiple sample splits which would require the adaption of the weight vectors. 
But we are currently implementing a way to pass predictions which would enable the use of e.g. weighted RandomForest

**Reply by gcasamat** [reply]
Thanks a lot. Look forward to this new feature.

**Reply by SvenKlaassen** [reply]
In the current development version external predictions are possible for (most) of the implemented models.
An example:

```
import numpy as np
import doubleml as dml
from doubleml.datasets import make_irm_data
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier

np.random.seed(3333)
data = make_irm_data(theta=0.5, n_obs=500, dim_x=20, return_type='DataFrame')
obj_dml_data = dml.DoubleMLData(data, 'y', 'd')

# DoubleML with interal predictions
ml_g = RandomForestRegressor(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
ml_m = RandomForestClassifier(n_estimators=100, max_features=20, max_depth=5, min_samples_leaf=2)
dml_irm_obj = dml.DoubleMLIRM(obj_dml_data, ml_g, ml_m)
dml_irm_obj.fit()
print(dml_irm_obj.summary)
```
External predictions should be a nested dictionary (outer level keys are the treatment names and inner level keys the corresponding nuisance elements). The values of the inner dictionary should have the shape `(n_obs, n_rep)`.

```
# predictions have the form (n_obs, n_rep, n_treatments)
pred_dict = {"d": {
    "ml_g0": dml_irm_obj.predictions["ml_g0"][:, :, 0],
    "ml_g1": dml_irm_obj.predictions["ml_g1"][:, :, 0],
    "ml_m": dml_irm_obj.predictions["ml_m"][:, :, 0]
    }               
}
```
The predictions can be supplied via the `external_predictions` argument in the `fit()` method

```
# define learners (not fitted for external predictions but needed for initialization)
ml_g = dml.utils.DMLDummyRegressor()
ml_m = dml.utils.DMLDummyClassifier()
dml_irm_obj_ext = dml.DoubleMLIRM(obj_dml_data, ml_g, ml_m)
dml_irm_obj_ext.fit(external_predictions=pred_dict)
print(dml_irm_obj_ext.summary)
```

Both models should give identical results. Remark that I used (new) dummy learners. The DoubleMLIRM class requires learners `ml_g` and `ml_m` for initialization but the dummy learners are not used. They raise errors if the are fitted etc. In principle this code also works with other learners but you wouldnt see if the learner is used or not (the dummy learners are just a safeguard against potential bugs).
I will comment if the version is released (and the documentation is updated).

**Reply by SvenKlaassen** [reply, marked as answer]
The option to use external predictions is available with version ``0.7.1``. You can find a documentation [here](https://docs.doubleml.org/dev/guide/learners.html#advanced-external-predictions).
